### ISSUE TITLE
User defined records and sumtypes

### ISSUE THREAD

**Issue by htngr** [nushell/nushell#11165]
### Related problem

See #11108 for the original motivation for this feature request.

### Describe the solution you'd like

A really useful feature of Rust and Haskell is *Algebraic Datatypes*, a fancy name for records and sum types. The latter is known as `enum` in Rust and is used for `Result`, `Option` and so on. This is one of the Features which set Rust and Haskell apart from more mainstream languages like Java, Go, C# and so on.

Nushell already has `record<...>` and `table<...>`, but lacks a convenient way to define them (see #10630). 
🎯 I think an ergonomic solution is to allow multiline record declarations and standalone types, just like `struct` in Rust:
```nushell
type foo = record<
    bar: int,
    baz: string,
    my_field: my_custom_type,
    my_other_field: my_alias
    ...
>
type my_custom_type = record<...>
type my_alias = decimal
```

On the other hand, Nushell lacks sum types. The syntax could look like this:
```nushell
type my_sumtype = enum<
  variant_a: int,
  variant_b: record<foo: string, bar: int>
>
```
Also they require pattern matching support to force handling *all* possible variants of a sum type:
```nushell
def handle [x: my_sumtype] {
  match $x {
    variant_a value => { print ($value * 2) }
    variant_b r => { print $"(r.foo): (r.bar)" }
  }
}
> handle (variant_a 21)
42
> handle (variant_b {foo: "Foo", bar: 123 })
Foo: 123
```

### Additional context and details

Some comments from @sholderbach:

> * named types   
>   * although pretty basic, it should be possible to add in backwards compatible ways
>   * discussion around the semantics of a nominal and a structural type system:
>   * Currently in Nushell structural `record<a: int, b: string> : record<a: int>`
> * enums, sum types   
>   * already have the notion of `match`     
>     * different exhaustiveness semantics?
>   * possible evolution: simpler pure nominal enums -> tagged union
>   * how would commands or cell path access play with tagged union type? as long as existing behavior can be preserved less of a backwards compatibility risk
> * user definable generics   
>   * should purely add expressive power and not impact existing concrete types
>   * existing precedent that adding this later can work (e.g. C# 1.0 -> 2.0, verdict still out on Go)

**Comment by htngr** [comment]
> > - discussion around the semantics of a nominal and a structural type system:
> > - Currently in Nushell structural record<a: int, b: string> : record<a: int>

I'm not very deep into this, but over the years I got the impression that subtyping and functional programming don't play well together (see for example <https://stackoverflow.com/questions/7266596/why-avoid-subtyping>). In OO for example this leads to the whole covariance / contravariance thing which can be very limiting. 
If I understand this correctly, the need for subtyping and structural types (like in the example `record<a: int, b: string> : record<a: int>`) is removed by *row-polymorphism* as implemented by OCaml, Elm and PureScript (also structural typing does hide information, while row polymorphism doesn't: <https://dev.to/maxheiber/subtyping-loses-information-row-polymorphism-does-not-2mb9>).

**Comment by amtoine** [comment]
yeah that would at least make long type annotations much easier to work with :pray:

**Comment by bew** [comment]
Hello!

Has there been any progress on the topic of declaring user-defined records?
(aka type alias to re-use a potentially long record type across multiple function definitions / vars / ..)

👉 I think it would be a good small first step with huge benefits for scripts wanting to have good signatures (e.g. for LLMs), without having looong lines nor having to repeat them at every type notations 🤔

WDYT?
